### PR TITLE
 Add data model for DTE Referencia

### PR DESCRIPTION
**Pull request by yaselc**
- Add a new enum for 'Tipo de Documento'
- Add `DteXmlReferencia` class.
- Update DTE's parser
- Add tests

Ref: https://cordada.aha.io/features/COMPCLDATA-221
Resolves: https://github.com/fyntex/lib-cl-sii-python/issues/217

### CODE REVIEW

File: cl_sii/dte/constants.py [hunk 1]
```diff
@@ -6,6 +6,187 @@
 
 """
 import enum
+from datetime import date
+from typing import FrozenSet
+
+
+@enum.unique
+class TipoDocumentoEnum(enum.IntEnum):
```
Comment (glarrain) [inline cl_sii/dte/constants.py:14]: Please see `RcvTipoDocto` https://github.com/fyntex/lib-cl-sii-python/blob/6d84a11b235f9302516a7972347a33178b6a9373/cl_sii/rcv/constants.py#L101
At first sight it seems to be pretty similar so a merge/refactor might be in order.

Also, since by "Documento" here we mean something that might not be a DTE necessarily, I think it should in another module. Perhaps in `cls_sii.base.constants` ? I'm not sure where.

File: cl_sii/dte/constants.py [hunk 2]
```diff
@@ -388,3 +390,193 @@ def emisor_is_vendedor(self) -> bool:
...
+#     (https://github.com/cl-sii-extraoficial/archivos-oficiales/blob/master/src/code/rtc/2019-12-12-schema_cesion/schema_cesion/DTE_v10.xsd#L1814-L1837)
+# TODO: Use `TipoDocumentoEnum` after deprecate `TipoDteEnum`
+DTE_REFERENCIA_CODREF_TIPO_DOC_MANDATORY_SET: FrozenSet[TipoDteEnum] = frozenset({
+    TipoDteEnum.NOTA_CREDITO_ELECTRONICA,
+    TipoDteEnum.NOTA_DEBITO_ELECTRONICA,
+})
+
+
+@enum.unique
+class CodigoReferenciaEnum(enum.IntEnum):
```
Comment (glarrain) [inline cl_sii/dte/constants.py:528]: Please remove the suffix `Enum` from `CodigoReferenciaEnum`. Perhaps you were inspired by `TipoDteEnum` but that one, authored by me, shouldn't have had the suffix as well.

File: cl_sii/dte/data_models.py [hunk 3]
```diff
@@ -466,6 +468,168 @@ def as_dte_data_l1(self) -> DteDataL1:
...
+    razon_ref: Optional[str] = None
+    """
+    The reason the document is being referenced
+
+    > Razon Explicita por la que se Referencia el Documento
+
+    DTE doc XML element: '..//Documento//Referencia//RazonRef'
+    """
+
+    def __post_init__(self) -> None:
```
Comment (glarrain) [inline cl_sii/dte/data_models.py:578]: Since this is a Pydantic dataclass, we might as well leverage Pydantic's validators. @jtrh suggestions?

File: cl_sii/dte/data_models.py [hunk 4]
```diff
@@ -630,7 +630,12 @@ def __post_init__(self) -> None:
                 self.razon_ref)
 
 
-@dataclasses.dataclass(frozen=True)
+@pydantic.dataclasses.dataclass(
```
Comment (glarrain) [inline cl_sii/dte/data_models.py:633]: Please mention in the commit message that this dataclass is now a Pydantic dataclass (instead of a std-lib one)

Comment (glarrain) [reply cl_sii/dte/data_models.py:633]: Also, since this dataclass is now a Pydantic dataclass, I think `dc_field` shouldn't be used anymore. @jtrh ?

File: cl_sii/dte/parse.py [hunk 5]
```diff
@@ -512,6 +530,91 @@ def _text_strip_or_raise(xml_em: XmlElement) -> str:
...
+def _validate_rut(v: object) -> object:
+    """
+    Reusable Pydantic validator for fields of type :class:`Rut`.
+    """
+    if isinstance(v, str):
+        v = Rut(value=v, validate_dv=False)  # Raises ValueError if invalid.
+    return v
+
+
+class _Referencia(pydantic.BaseModel):
```
Comment (glarrain) [inline cl_sii/dte/parse.py:546]: Shouldn't this be named something like `_DteXmlReferenciaParser` ?

File: cl_sii/dte/parse.py [hunk 6]
```diff
@@ -512,6 +530,91 @@ def _text_strip_or_raise(xml_em: XmlElement) -> str:
...
+            fecha_ref=xml_em.findtext('sii-dte:FchRef', namespaces=DTE_XMLNS_MAP),
+            codigo_ref=xml_em.findtext('sii-dte:CodRef', namespaces=DTE_XMLNS_MAP) or None,
+            razon_ref=xml_em.findtext('sii-dte:RazonRef', namespaces=DTE_XMLNS_MAP) or None,
+        )
+
+    ###########################################################################
+    # Validators
+    ###########################################################################
+
+    _validate_rut = pydantic.validator(  # type: ignore[pydantic-field]
```
Comment (glarrain) [inline cl_sii/dte/parse.py:613]: I'm not familiar with this pattern. Are you sure it is used like this? Is it possible to change something so we don't need the  `# type: ignore`?

File: tests/test_data/sii-dte/DTE--96670340-7--61--110616--cleaned.xml [hunk 7]
```diff
@@ -0,0 +1,158 @@
+<?xml version='1.0' encoding='ISO-8859-1'?>
```
Comment (glarrain) [inline tests/test_data/sii-dte/DTE--96670340-7--61--110616--cleaned.xml:1]: If this is the cleaned file, shouldn't it have fixed the missing newline at the end? I don't remember what we did for the other "cleaned" files

File: cl_sii/dte/data_models.py (hunk 4 above)
Comment (jtrobles-cdd) [reply cl_sii/dte/data_models.py:633]: > Also, since this dataclass is now a Pydantic dataclass, I think `dc_field` shouldn't be used anymore. @jtrh ?

`dc_field` (`dataclasses.field`) can be used both with Standard-Library Dataclasses and Pydantic Dataclasses. However, all the current uses of `dc_field` in `DteXmlData` are _already_ unnecessary, as `dc_field` is only required for some special cases, such as mutable default values (e.g. `mylist: List[int] = dc_field(default_factory=list)` instead of `mylist: List[int] = []`).

` = dc_field()` can be safely deleted, and ` = dc_field(default=None)` can be replaced with `= None`.

Related documentation:

- https://docs.python.org/3.7/library/dataclasses.html
- https://docs.python.org/3.7/library/dataclasses.html#dataclasses.field
- https://docs.python.org/3.7/library/dataclasses.html#init-only-variables

File: cl_sii/dte/parse.py (hunk 6 above)
Comment (jtrobles-cdd) [reply cl_sii/dte/parse.py:613]: > I'm not familiar with this pattern.

It's a [reusable Pydantic validator](https://pydantic-docs.helpmanual.io/usage/validators/#reuse-validators).

File: cl_sii/dte/constants.py (hunk 1 above)
Comment (jtrobles-cdd) [inline cl_sii/dte/constants.py:14]: - Evaluate merging with `cl_sii.dte.constants.TipoDocumentoEnum` and/or `cl_sii.rcv.constants.RcvTipoDocto`. That would be a big change, so I recommend doing it in a separate pull request.
- Remove `Enum` suffix from class name.
- Consider moving it to `cl_sii.base.constants` because it looks like it could be useful across multiple SII subpackages.

File: cl_sii/dte/data_models.py (hunk 3 above)
Comment (jtrobles-cdd) [reply cl_sii/dte/data_models.py:578]: > [...] leverage Pydantic's validators. @jtrh suggestions?

- Refactor all validations done in `__post_init__` as Pydantic validators. Example: https://github.com/fyntex/lib-cl-sii-python/blob/fd39c62d2c7fedfd7e044bfbbab3a0f19531e6db/cl_sii/rtc/data_models.py#L37-L47 and https://github.com/fyntex/lib-cl-sii-python/blob/fd39c62d2c7fedfd7e044bfbbab3a0f19531e6db/cl_sii/rtc/data_models.py#L388-L392.
- Type checks using `isinstance` are not needed because they're automatically performed by Pydantic.

File: cl_sii/dte/data_models.py [hunk 8]
```diff
@@ -466,7 +468,174 @@ def as_dte_data_l1(self) -> DteDataL1:
...
+                self.fecha_ref)
+
+        if self.razon_ref and len(self.razon_ref) > constants.DTE_REFERENCIA_RAZON_MAX_LENGTH:
+            raise ValueError(
+                "The maximum length allowed for `razon_ref` is "
+                f"{constants.DTE_REFERENCIA_RAZON_MAX_LENGTH}",
+                self.razon_ref)
+
+
+@pydantic.dataclasses.dataclass(
```
Comment (jtrobles-cdd) [inline cl_sii/dte/data_models.py:633]: I would move the conversion of `DteXmlData` to a Pydantic Dataclass into a separate commit and/or pull request.

File: cl_sii/dte/data_models.py [hunk 9]
```diff
@@ -495,32 +664,32 @@ class DteXmlData(DteDataL1):
     # fields
     ###########################################################################
 
-    emisor_razon_social: str = dc_field()
+    emisor_razon_social: str
```
Comment (jtrobles-cdd) [inline cl_sii/dte/data_models.py:667]: I would move this, and similar changes, into a separate commit and/or pull request.

File: cl_sii/dte/data_models.py [hunk 10]
```diff
@@ -529,21 +698,29 @@ class DteXmlData(DteDataL1):
...
 
+    referencias: Optional[Sequence[DteXmlReferencia]] = None
+    """
+    List of structs for ``Referencia`` XML elements.
+
+    ..warning::
+        The items MUST be ordered according to their ``numero_linea_ref``.
+    """
+
     def __post_init__(self) -> None:
```
Comment (jtrobles-cdd) [inline cl_sii/dte/data_models.py:724]: If `DteXmlData` is going to be converted into Pydantic Dataclass, then either move the validations performed in `__post_init__` to `__post_init_post_parse__`, or refactor them as Pydantic validators. Either way, please do it in a separate commit and/or pull request.

Pydantic's documentation says:

> When substituting usage of `dataclasses.dataclass` with `pydantic.dataclasses.dataclass`, it is recommended to move the code executed in the `__post_init__` method to the `__post_init_post_parse__` method, and only leave behind part of code which needs to be executed before validation.

Source: https://pydantic-docs.helpmanual.io/usage/dataclasses/#difference-with-stdlib-dataclasses

File: cl_sii/dte/parse.py [hunk 11]
```diff
@@ -538,6 +556,93 @@ def _text_strip_or_raise(xml_em: XmlElement) -> str:
...
```
Comment (fpinto-cdd) [inline cl_sii/dte/parse.py:639]: This `# type: ignore` is no longer needed

```suggestion
    _validate_rut = pydantic.validator(
```

File: tests/test_dte_data_models.py [hunk 12]
```diff
@@ -18,12 +18,13 @@
...
     DteNaturalKey,
     DteXmlData,
+    DteXmlReferencia,
     validate_contribuyente_razon_social,
     validate_dte_folio,
     validate_dte_monto_total,
 )
 from cl_sii.libs import encoding_utils, tz_utils
-from cl_sii.rut import Rut
+from cl_sii.rut import Rut  # noqa: F401
```
Comment (fpinto-cdd) [inline tests/test_dte_data_models.py:27]: This `noqa` is no longer needed 

```suggestion
from cl_sii.rut import Rut
```

**Review by fpinto-cdd — COMMENTED** [review]
- Let's change the title of this commit to "chore: Add Pydantic validations to data model `DteXmlReferencia`".
- Let's change the marked word `classes` to `validators`.

Captura de pantalla 2023-03-22 173606

**Review by fpinto-cdd — APPROVED** [review]

File: cl_sii/base/constants.py
Comment (jtrobles-cdd) [inline cl_sii/base/constants.py:1]: What do you mean by *"at `sii_dte`*" in the commit title?

Comment (jtrobles-cdd) [inline cl_sii/base/constants.py:1]: In the commit title, can you replace the `chore:` prefix with `feat(base)`?

**Review by jtrobles-cdd — CHANGES_REQUESTED** [review]

**Review by jtrobles-cdd — CHANGES_REQUESTED** [review]
- [ ] Can you change the commit type of _"chore: Add Pydantic validators to data model `DteXmlReferencia`"_ from `chore` to `feat(dte)`?
- [ ] In commit _"feat(base): Add enum for 'Tipo de Documento'"_, can you use permalinks (with 8-character shortened hash) for the GitHub URLs?

**Review by jtrobles-cdd — APPROVED** [review]
